File: examples/C++/MyProjects/Doodle/step05/DrawWindow.cc
// DrawWindow.cc: implements DrawWindow class
#include "DrawWindow.h"
#include "Line.h"
#include "chocolaf.h"
#include <QApplication>
#include <QColorDialog>
#include <QInputDialog>
#include <QMessageBox>
#include <QtGui>

const QString AppTitle("Qt Scribble");
const QString WindowTitle =
    QString("Qt %1 Doodle - Step05: Changing Line color & thickness")
        .arg(QT_VERSION_STR);

DrawWindow::DrawWindow() {
  setAttribute(Qt::WA_StaticContents);
  setWindowTitle(WindowTitle);
  _modified = false;
  _dragging = false;
  _penWidth = 3;
  _penColor = QColor("#ff557f"); // qRgb(0, 0, 255);
  _line = nullptr;
}

DrawWindow::~DrawWindow() { delete _line; }

void DrawWindow::closeEvent(QCloseEvent *event) {
  // window is about to close, prompt user & decide if ok to quit
  // based on user's response.
  if (_modified) {
    switch (QMessageBox::question(this, tr("Qt Scribble Tutorial"),
                                  tr("The doodle has been modified.\nDo you "
                                     "want to close without saving?"),
                                  QMessageBox::Yes | QMessageBox::No,
                                  QMessageBox::No)) {
    case QMessageBox::Yes:
      // ok to quit
      qDebug() << "User chose to quit without saving!";
      event->accept();
      break;
    default:
      // don't quit yet
      event->ignore();
    }
  }
}

void DrawWindow::drawLineTo(const QPoint &pt) {
  // draw line from _lastPt to pt
  QPainter painter(&_image);
  QPen pen(_penColor, _penWidth);
  painter.setRenderHint(QPainter::Antialiasing);
  painter.setPen(pen);
  painter.drawLine(_lastPt, pt);
  _lastPt = pt;
  update();
}

void DrawWindow::clearImage() {
  //_image.fill(qRgb(255, 255, 255));
  _image.fill(Chocolaf::ChocolafPalette::Window_Color);
  update();
}

void DrawWindow::mousePressEvent(QMouseEvent *event) {
  if (event->button() == Qt::LeftButton) {
    // left mouse button pressed
    if (QApplication::keyboardModifiers() & Qt::ControlModifier) {
      // display message box & get new width of pen
      // NOTE: this is clearly an ugly UI design. Please excuse for now
      bool ok;
      int newPenWidth = QInputDialog::getInt(
          this, AppTitle, QString("Enter new pen width (2-12):"),
          _line->penWidth(), 2, 12, 1, &ok);
      if (ok) {
        qDebug() << "New pen width selected: " << newPenWidth;
        _penWidth = newPenWidth;
      }
    } else {
      delete _line;
      _line = nullptr;
      clearImage();
      _lastPt = event->pos();
      _dragging = true;
      _modified = true;
      _line = new Line(_penWidth, _penColor);
      _line->addPoint(event->pos());
    }
  } else if (event->button() == Qt::RightButton) {
    if (QApplication::keyboardModifiers() & Qt::ControlModifier) {
      // if Ctrl key is also pressed, display standard color
      // dialog & get new pen color
      QColor color = QColorDialog::getColor(_penColor, this);
      if (color.isValid())
        _penColor = color;
    } else {
      clearImage();
      _modified = false;
    }
  }
}

void DrawWindow::mouseMoveEvent(QMouseEvent *event) {
  if (event->buttons() & Qt::LeftButton && _dragging) {
    drawLineTo(event->pos());
    _line->addPoint(event->pos());
  }
}

void DrawWindow::mouseReleaseEvent(QMouseEvent *event) {
  if ((event->button() == Qt::LeftButton) && _dragging) {
    drawLineTo(event->pos());
    _line->addPoint(event->pos());
    _dragging = false;
  }
}

void DrawWindow::resizeEvent(QResizeEvent *event) {
  if (width() > _image.width() || height() > _image.height()) {
    int newWidth = qMax(width(), _image.width());
    int newHeight = qMax(height(), _image.height());
    resizeImage(QSize(newWidth, newHeight));
    update();
  }
  QWidget::resizeEvent(event);
}

void DrawWindow::paintEvent(QPaintEvent *event) {
  QPainter painter(this);
  painter.setRenderHint(QPainter::Antialiasing);
  QRect dirtyRect = event->rect();
  painter.drawImage(dirtyRect, _image, dirtyRect);
}

void DrawWindow::resizeImage(const QSize &newSize) {
  if (_image.size() == newSize)
    return;
  QImage newImage(newSize, QImage::Format_RGB32);
  // newImage.fill(qRgb(255, 255, 255));
  newImage.fill(Chocolaf::ChocolafPalette::Window_Color);

  // draw existing image over new image
  QPainter painter(&newImage);
  painter.setRenderHint(QPainter::Antialiasing);
  painter.drawImage(QPoint(0, 0), _image);
  _image = newImage;
}

File: examples/Python/MyProjects/PyDoodle/step07/drawWindow.py
# -*- coding: utf-8 -*-
"""
// ============================================================================
// drawWindow.py: custom QMainWindow derived class for main window
//
// Tutorial - PyQt5 Doodle Application
// Based on a similar tutorial for Borland ObjectWindows Library (OWL)
// @author: Manish Bhobe
// My experiments with the Qt Framework. Use at your own risk!!
// ============================================================================
"""

from PyQt6.QtCore import *
from PyQt6.QtGui import *
from PyQt6.QtWidgets import *

from chocolaf import ChocolafPalette
from doodle import Doodle
from squiggle import Squiggle


class DrawWindow(QWidget):
    def __init__(self, *args, **kwargs):
        super(QWidget, self).__init__(*args, **kwargs)
        self.setStyleSheet("background-color: white")
        self.__doodle = Doodle()
        self.__dragging = False
        self.__currSquiggle = None
        self.setMouseTracking(True)
        self._timerID = -1  # self.startTimer(1000)

    @property
    def doodle(self):
        return self.__doodle

    def timerEvent(self, e: QTimerEvent) -> None:
        if e.timerId() == self._timerID:
            if darkdetect.isDark():
                utils.setDarkPalette(qApp)
            else:
                utils.setDarkPalette(qApp, False)

    def paintEvent(self, e: QPaintEvent) -> None:
        """handler for paint events"""
        painter = QPainter()
        try:
            width, height = self.width(), self.height()
            color = ChocolafPalette.Window_Color
            painter.begin(self)
            painter.setBrush(color)
            painter.setPen(color)
            painter.drawRect(0, 0, width, height)
            painter.setRenderHint(QPainter.Antialiasing, True)
            self.doodle.draw(painter)
        finally:
            painter.end()

    def mousePressEvent(self, e: QMouseEvent) -> None:
        """handler for mouse press (left or right clostBtn) events"""
        if e.button() == Qt.LeftButton:
            if e.modifiers() & Qt.ControlModifier:
                # if Ctrl key is also pressed with mouse press, display
                # dialog to change pen thickness
                newWidth, ok = QInputDialog.getInt(
                    self, "Pen Width", "Enter new pen width (2-12):", self.doodle.defPenWidth, 2, 12
                )
                if ok:  # user clicked Ok on QInputDialog
                    self.doodle.defPenWidth = newWidth
            else:
                # start a new line
                # self.__currSquiggle = Squiggle(self.penWidth, self.penColor)
                self.__currSquiggle = Squiggle()
                self.__currSquiggle.penWidth = self.doodle.defPenWidth
                self.__currSquiggle.penColor = self.doodle.defPenColor
                self.doodle.append(self.__currSquiggle)
                pt = QPoint(e.pos().x(), e.pos().y())
                self.__currSquiggle.append(pt)
                self.doodle.modified = True
                self.__dragging = True
        elif e.button() == Qt.RightButton:
            if e.modifiers() & Qt.ControlModifier:
                # if Ctrl key is also pressed with mouse press, display
                # dialog to change pen color
                newColor = QColorDialog.getColor(self.doodle.defPenColor, self)
                if newColor.isValid():
                    self.doodle.defPenColor = newColor
            else:
                self.doodle.clear()
                self.update()

    def mouseMoveEvent(self, e: QMouseEvent) -> None:
        """handler for mouse drag (left or right clostBtn) events
        NOTE: you must call setMouseTracking(True) so window can receive mouse drag events
        """
        if (e.buttons() == Qt.LeftButton) and (self.__dragging):
            assert self.__currSquiggle != None, "FATAL: self.currLine is None, when expecting valid!"
            pt = QPoint(e.pos().x(), e.pos().y())
            self.__currSquiggle.append(pt)
            self.update()
        else:
            e.accept()

    def mouseReleaseEvent(self, e: QMouseEvent) -> None:
        """handler for mouse (left or right clostBtn) released events"""
        if (e.button() == Qt.LeftButton) and (self.__dragging):
            assert self.__currSquiggle != None, "FATAL: self.currLine is None, when expecting valid!"
            pt = QPoint(e.pos().x(), e.pos().y())
            self.__currSquiggle.append(pt)
            self.__dragging = False
            self.update()
        else:
            e.accept()
